hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -1768,7 +1768,7 @@ object PruneDeadFields {
         case ta: TStreamable =>
           val ra = rType.asInstanceOf[TStreamable]
           val uid = genUID()
-          val ref = Ref(uid, -ta.elementType)
+          val ref = Ref(uid, ta.elementType)
```

Comment:
Because we shouldn't be relying on requiredness, which also means not explicitly flipping required = off. PruneDeadFields.upcast now ignores requiredeness (checking `ir.typ isOfType rType`), as does InferType.

Reply:
> PruneDeadFields.upcast now ignores requiredeness

I don't think you can do this yet.

Reply:
If, for instance, you rebuild a TableUnion's children to have different requiredness in their virtual types, you **must** upcast in a way that changes requiredness.

Reply:
Ah, right, without this change to PruneDeadFields, MakeArray/Coalesce/If will *never* get anything other than optional types in Emit, which means the Emit test tests nothing.

Reply:
I don't think I understand. If you have:
```scala
If(..., In(0, TArray(TInt32(true)), In(1, TArray(TInt32(false)))
```

This should be inferred as type `TArray(TInt32(false))`, and the emit test will test that the the consequent is correctly converted.

Reply:
Here is the test case

```scala
@Test def testMakeArrayWithDifferentRequiredness(): Unit = {
    val t = TArray(TStruct("a" -> TInt32Required, "b" -> TArray(TInt32Optional, required = true)))
    val value = Row(2, FastIndexedSeq(1))
    assertEvalsTo(
      MakeArray(
        Seq(In(0, PStruct("a" -> PInt32(), "b" -> PArray(PInt32()))), In(1, t.elementType)),
        t.deepOptional().asInstanceOf[TArray]
      ),
      FastIndexedSeq((null, t.elementType.deepOptional()), (value, t.elementType)),
      FastIndexedSeq(null, value)
    )
  }
```

Here is the Emit code:

```scala
case x@MakeArray(args, _) =>
       args.map(x => println(s"ARG ${x.pType}"))
       # ... more
```

With `==` in upcast (`if(ir.typ == rType) ir else { ... }`):
```
ARG PCStruct{a:PInt32,b:PCArray[PInt32]}
ARG PCStruct{a:PInt32,b:PCArray[PInt32]}
```

With `isOfType` in upcast (`if(ir.typ isOfType rType) ir else { ... }`):

```
ARG PCStruct{a:PInt32,b:PCArray[PInt32]}
ARG PCStruct{a:+PInt32,b:+PCArray[PInt32]}
```

Because upcast triggers

Reply:
You should change the prune rebuild rule. If there is at least one child IR and all others compare `isOfType` true against that IR, then don't call upcast.

Reply:
Bah, why:

```scala
case MakeArray(args, _) =>
        val args2 = args.map(rebuildIR(_, env, memo))
        if(args2.forall(_.typ.isOfType(args2.head.typ)))
          MakeArray(args2, args2.head.typ.asInstanceOf[TArray])
        else {
          val depArray = requestedType.asInstanceOf[TArray]
          MakeArray(args2.map(a => upcast(a, depArray.elementType)), depArray)
        }
````

"""
2020-01-27 17:32:28 Hail: ERROR: error from strategy Interpret



is.hail.utils.HailException: error trying to rebuild IR:
(MakeArray Array[Struct{a:Int32,b:Array[Int32]}]
  (In PCStruct{a:PInt32,b:PCArray[PInt32]} 0)
  (In PCStruct{a:PInt32,b:PCArray[PInt32]} 1))
""""

Reply:
`args2.head.typ.asInstanceOf[TArray]` is wrong. Need to match what InferType would give you, I think.